### PR TITLE
Short URLs

### LINKED ISSUE #369: Restructure repository functions
**Is your feature request related to a problem? Please describe.**
Currently, we have a massive single file for data retrieval and manipulation of all models in repo.py.

**Describe the solution you'd like**
Make repo a directory and provide one python module per model. Simplify function names through scope implication (e.g. `repo.delete_subscriber()` -> `repo.subscriber.delete()` instead of `repo.subscriber.delete_subscriber()`).

### PR DESCRIPTION
**Pull request by MelissaAutumn**
Fixes #369 

This PR introduces schedule slugs and the ability to view availability using a schedule slug.

We'll phase out signed urls for the booking page eventually, and the user currently cannot customize a schedule slug, but this is a first step into making the "my link" small. I've also increased the space the "my link" textbox has by moving the copy button to a tooltip.

(Note: I have short links disabled on local dev, so on stage/prod it won't show the /user/ part.)
image

When you refresh your link, all schedule slugs will also be refreshed. Basically we shouldn't even see the booking signed urls anymore.

### CODE REVIEW

File: backend/src/appointment/database/models.py [hunk 1]
```diff
@@ -80,9 +81,15 @@ class InviteStatus(enum.Enum):
     revoked = 2  # The code is no longer valid and cannot be used for sign up anymore
 
 
+def encrypted_type(column_type, length: int = 255, **kwargs) -> StringEncryptedType:
```
Comment (MelissaAutumn) [inline backend/src/appointment/database/models.py:84]: I snuck this in to reduce visual noise. We can eventually add some functionality to not encrypt local dev environments for better debugging.

File: backend/src/appointment/database/models.py [hunk 2]
```diff
@@ -236,38 +245,49 @@ class Schedule(Base):
...
     @property
     def end_time_local(self) -> datetime.time:
         """End Time in the Schedule's Calendar's Owner's timezone"""
-        time_of_save = self.time_updated.replace(hour=self.end_time.hour, minute=self.end_time.minute, second=0, tzinfo=datetime.timezone.utc)
+        time_of_save = self.time_updated.replace(hour=self.end_time.hour, minute=self.end_time.minute, second=0,
+                                                 tzinfo=datetime.timezone.utc)
         return time_of_save.astimezone(zoneinfo.ZoneInfo(self.calendar.owner.timezone)).time()
 
+    @cached_property
+    def owner(self):
```
Comment (MelissaAutumn) [inline backend/src/appointment/database/models.py:286]: Convenience property

File: backend/src/appointment/database/repo/schedule.py [hunk 3]
```diff
@@ -69,3 +75,36 @@ def has_slot(db: Session, schedule_id: int, slot_id: int):
...
+
+def generate_slug(db: Session, schedule_id: int) -> str|None:
+    schedule = repo.schedule.get(db, schedule_id)
+
+    if schedule.slug:
+        return schedule.slug
+
+    # If slug isn't provided, give them the last 8 characters from a uuid4
+    # Try up-to-3 times to create a unique slug
+    for _ in range(3):
```
Comment (MelissaAutumn) [inline backend/src/appointment/database/repo/schedule.py:88]: I'm uhh not sure on a better way to do this right now.

File: backend/src/appointment/routes/api.py [hunk 4]
```diff
@@ -101,6 +101,16 @@ def refresh_signature(db: Session = Depends(get_db), subscriber: Subscriber = De
         subscriber.id,
     )
 
+    # Update schedule slugs as well!
+    # This is temp until we figure this flow out
+    schedules = repo.schedule.get_by_subscriber(db, subscriber.id)
```
Comment (MelissaAutumn) [inline backend/src/appointment/routes/api.py:106]: Once we have multiple schedules we'll have to figure out a better flow for showing personal links and such. I assume we'll have one "General availability" schedule, but that's for future us!

File: frontend/src/views/BookingView.vue [hunk 5]
```diff
@@ -130,24 +132,43 @@ const getViewBySlotDistribution = (slots) => {
...
 /**
  * Retrieve the appointment from either availability or booking routes.
  * Returns null if there was an error, or the Appointment object if it was successful.
  * @returns {Promise<Object|null>}
  */
 const getAppointment = async () => {
-  const request = call('schedule/public/availability').post({ url: window.location.href.split('#')[0] });
+  let url = null;
+  // Okay we have a slug, lets lookup the actual signature
+  if (route.params.usernameOrSlug && !route.params.signature) {
```
Comment (MelissaAutumn) [inline frontend/src/views/BookingView.vue:153]: If come in with a slug, we look it up in on the server to retrieve their signed url, and then use that to fetch. 

This is a staged approach, and we can easily remove the user's signedUrl, and this intermediate step in the future.

**Review by devmount — APPROVED** [review]
Looking good so far!

File: backend/src/appointment/database/repo/schedule.py (hunk 3 above)
Comment (MelissaAutumn) [reply backend/src/appointment/database/repo/schedule.py:95]: Yeah we always want a upper limit on these types of loop. So 3 is the limit here 😄 

They're also namespaced to the subscriber so it's very unlikely it'll ever create a duplicate.

**Comment by MelissaAutumn** [comment]
I still need to rebase the linting changes, but it's ready for review now.

**Review by devmount — COMMENTED** [review]
Hm, the second migration file line 22 throws an error on my end. Seems to be cused by the None check `models.Schedule.slug.is_(None)`. Or do you have any ideas Mel?

```
root@115f5bc69013:/app# alembic upgrade head
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade d5de8f10ab87 -> f1e20604d6e8, add slug to schedules
INFO  [alembic.runtime.migration] Running upgrade f1e20604d6e8 -> 12c7e1b34dd6, fill empty schedule slugs
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/cryptography/utils.py", line 35, in _check_byteslike
    memoryview(value)
TypeError: memoryview: a bytes-like object is required, not 'NoneType'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/alembic/config.py", line 590, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.11/site-packages/alembic/config.py", line 584, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.11/site-packages/alembic/config.py", line 561, in run_cmd
    fn(
  File "/usr/local/lib/python3.11/site-packages/alembic/command.py", line 378, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.11/site-packages/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/app/appointment/migrations/env.py", line 102, in <module>
    run_migrations_online()
  File "/app/appointment/migrations/env.py", line 96, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/usr/local/lib/python3.11/site-packages/alembic/runtime/environment.py", line 867, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/usr/local/lib/python3.11/site-packages/alembic/runtime/migration.py", line 624, in run_migrations
    step.migration_fn(**kw)
  File "/app/appointment/migrations/versions/2024_06_05_2036-12c7e1b34dd6_fill_empty_schedule_slugs.py", line 22, in upgrade
    schedules: list[models.Schedule] = session.query(models.Schedule).where(models.Schedule.slug.is_(None)).all()
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/query.py", line 2772, in all
    return self._iter().all()
           ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 1430, in all
    return self._allrows()
           ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 401, in _allrows
    rows = self._fetchall_impl()
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 1343, in _fetchall_impl
    return self._real_result._fetchall_impl()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 1755, in _fetchall_impl
    return list(self.iterator)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/loading.py", line 147, in chunks
    fetch = cursor._raw_all_rows()
            ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 393, in _raw_all_rows
    return [make_row(row) for row in rows]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/engine/result.py", line 393, in <listcomp>
    return [make_row(row) for row in rows]
            ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/sql/type_api.py", line 1710, in process
    return process_value(value, dialect)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy_utils/types/encrypted/encrypted_type.py", line 423, in process_result_value
    self._update_key()
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy_utils/types/encrypted/encrypted_type.py", line 392, in _update_key
    self.engine._update_key(key)
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy_utils/types/encrypted/encrypted_type.py", line 52, in _update_key
    digest.update(key)
  File "/usr/local/lib/python3.11/site-packages/cryptography/hazmat/primitives/hashes.py", line 98, in update
    utils._check_byteslike("data", data)
  File "/usr/local/lib/python3.11/site-packages/cryptography/utils.py", line 37, in _check_byteslike
    raise TypeError("{} must be bytes-like".format(name))
TypeError: data must be bytes-like
root@115f5bc69013:/app#
```

File: backend/src/appointment/database/repo/schedule.py (hunk 3 above)
Comment (devmount) [reply backend/src/appointment/database/repo/schedule.py:95]: Agreed. If the 3 attempts are not sufficient, we should start asking general questions about randomness 😂

File: backend/src/appointment/database/repo/subscriber.py [hunk 6]
```diff
@@ -1,13 +1,14 @@
 """Module: repo.subscriber
 
-Repository providing CRUD functions for subscriber database models.
+Repository providing CRUD functions for subscriber database models.
```
Comment (devmount) [inline backend/src/appointment/database/repo/subscriber.py:3]: Is this a linter issue?

File: backend/src/appointment/database/repo/subscriber.py [hunk 7]
```diff
@@ -122,28 +123,13 @@ def verify_link(db: Session, url: str):
...
-    # Flatten
-    match = match[0]
-    clean_url = url
-
-    username = match[0]
-    signature = None
-    if len(match) > 1:
-        signature = match[1]
-        clean_url = clean_url.replace(signature, '')
+    username, signature, clean_url = utils.retrieve_user_url_data(url)
```
Comment (devmount) [inline backend/src/appointment/database/repo/subscriber.py:126]: Yes, thank you for moving that to utils

**Review by devmount — APPROVED** [review]
Works! Thanks!

File: backend/src/appointment/routes/schedule.py [hunk 8]
```diff
@@ -275,7 +302,8 @@ def request_schedule_availability_slot(
 
     # Sending confirmation email to owner
     background_tasks.add_task(
-        send_confirmation_email, url=url, attendee_name=attendee.name, date=date, to=subscriber.preferred_email
+        send_confirmation_email, url=url, attendee_name=attendee.name, attendee_email=attendee.email, date=date,
+        to=subscriber.preferred_email
```
Comment (devmount) [inline backend/src/appointment/routes/schedule.py:306]: This is not really related to this PR, but I found this during testing and decided to just fix it. Somehow we missed providing the attendee email which resulted in an error which was thrown on booking.

File: backend/src/appointment/database/repo/subscriber.py (hunk 6 above)
Comment (MelissaAutumn) [reply backend/src/appointment/database/repo/subscriber.py:3]: Possibly my cat's doing 🤔

File: backend/src/appointment/routes/schedule.py (hunk 8 above)
Comment (MelissaAutumn) [reply backend/src/appointment/routes/schedule.py:306]: Thank you!
